zerver/lib/stream_subscription.py

```diff
@@ -274,3 +275,60 @@ def subscriber_ids_with_stream_history_access(stream: Stream) -> Set[int]:
             stream.id, include_deactivated_users=False
         ).values_list("user_profile_id", flat=True)
     )
+
+
+def get_subscriptions_for_send_message(
+    *,
+    stream_id: int,
+    possible_wildcard_mention: bool,
+    possibly_mentioned_user_ids: AbstractSet[int],
+) -> QuerySet:
+    """
+    Note that this code optimizes an important use
+    case for large streams. Certain realms have
+    many long-term-idle users that are subscribed
+    to large streams (usually default streams), and
+    we don't acually have to send messages to most
+    of the users. It's expensive to process them
+    in a naive query, hence this function.
+
+    This returns a superset of the subscriptions that
+    will require action in the send-message codepath.
+
+    Basically, it excludes all long-term idle users
+    who don't have stream/push options set and who
+    also are not mentioned in the message.
+
+    For expediency, it also returns all users on the
+    stream who have ANY alert words, even if their
+    alert words aren't in the message. (We create
+    UserMessage rows for long-term-idle messages if
+    they have been alerted. Downstream logic will
+    filter out false alarms.)
+
+    We also get false alarms for certain messages
+    that **look** like they are mentioning a user
+    from a quick parse. Downstream logic will make
+    sure we filter out "falsely" mentioned users.
+    """
+
+    query = Subscription.objects.filter(
+        recipient__type=Recipient.STREAM,
+        recipient__type_id=stream_id,
+        active=True,
+        is_user_active=True,
+    )
+
+    if possible_wildcard_mention:
+        return query
+
+    query = query.filter(
+        Q(user_profile__long_term_idle=False)
+        | Q(push_notifications=True)
+        | (Q(push_notifications=None) & Q(user_profile__enable_stream_push_notifications=True))
+        | Q(email_notifications=True)
+        | (Q(email_notifications=None) & Q(user_profile__enable_stream_email_notifications=True))
+        | Q(user_profile_id__in=possibly_mentioned_user_ids)
+        | Exists(AlertWord.objects.filter(user_profile_id=OuterRef("user_profile_id")))
```

Comment:
Posted this in the stream, but can you look at how the query plan differs with a `, realm=realm` here?  I think providing postgres with that detail may be helpful in zulipchat.com type environments to ensure query plans never try to do an "every alert word on the server" table scan.

(The other option is to use the `alert_words_in_realm` cache)

Reply:
I am now passing down realm_id. If we get a good query plan here, my guess is that it's just as efficient for DB to cross-check AlertWord as to pass in a list of ids, particularly for datasets where the DB never needs to look at AlertWord for most user_id values.  It should be easy enough to move to the cache in a follow up commit.